File: orchestrator/hooks/parser.py
"""Pure functions that translate hook payloads into domain decisions."""

from typing import Any

from orchestrator.core.sessions import SessionStatus


class InvalidHookPayloadError(Exception):
    pass


def parse_notification(payload: dict[str, Any]) -> SessionStatus:
    if "message" not in payload:
        raise InvalidHookPayloadError("Notification payload missing 'message'")
    return SessionStatus.AWAITING_RESPONSE


def parse_stop(_payload: dict[str, Any]) -> SessionStatus:
    return SessionStatus.IDLE


def parse_pretooluse(payload: dict[str, Any]) -> str:
    tool = payload.get("tool_name")
    if not isinstance(tool, str) or not tool:
        raise InvalidHookPayloadError("PreToolUse payload missing 'tool_name'")
    return tool

File: internal/hooks/parser.go
package hooks

import (
	"errors"
	"fmt"
)

// ErrInvalidHookPayload signals that a claude hook payload is missing required
// fields or has wrong types. Hook routes return HTTP 422 in this case.
var ErrInvalidHookPayload = errors.New("invalid hook payload")

// Session status constants. These match Python's SessionStatus StrEnum values
// and the CHECK constraint in the sessions table migration.
const (
	StatusExecuting        = "executing"
	StatusAwaitingResponse = "awaiting_response"
	StatusIdle             = "idle"
	StatusError            = "error"
	StatusDone             = "done"
)

// ParseNotification: claude says it produced a notification → user is awaited.
func ParseNotification(payload map[string]any) (string, error) {
	if _, ok := payload["message"]; !ok {
		return "", fmt.Errorf("ParseNotification: missing 'message': %w", ErrInvalidHookPayload)
	}
	return StatusAwaitingResponse, nil
}

// ParseStop: claude finished its turn → idle.
func ParseStop(_ map[string]any) (string, error) {
	return StatusIdle, nil
}

// ParsePreToolUse: claude is about to invoke a tool. Returns the tool name.
func ParsePreToolUse(payload map[string]any) (string, error) {
	name, ok := payload["tool_name"].(string)
	if !ok || name == "" {
		return "", fmt.Errorf("ParsePreToolUse: missing or non-string 'tool_name': %w", ErrInvalidHookPayload)
	}
	return name, nil
}
